### PR TITLE
Add --debug flag and internal debug arguments

### PR DESCRIPTION
**Pull request by userzimmermann**
@sils @Makman2 

Instead of the ERROR output from https://github.com/coala/coala/issues/3322 you can now get this:

```
> coala --debug
Executing section Default...
PicklingError("Can't pickle <PycodestyleBear linter class (wrapping 'pycodestyle')>: attribute lookup PycodestyleBear on coalib.bearlib.abstractions.Linter failed",)
> c:\users\zimmermann\miniconda3\envs\coala\lib\multiprocessing\reduction.py(59)dump()
     57 def dump(obj, file, protocol=None):
     58     '''Replacement for pickle.dump() using ForkingPickler.'''
---> 59     ForkingPickler(file, protocol).dump(obj)
     60
     61 #

ipdb> Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\Zimmermann\Miniconda3\envs\coala\lib\multiprocessing\spawn.py", line 106, in spawn_main
    exitcode = _main(fd)
  File "C:\Users\Zimmermann\Miniconda3\envs\coala\lib\multiprocessing\spawn.py", line 116, in _main
    self = pickle.load(from_parent)
EOFError: Ran out of input
ipdb>
```

And by calling `coalib.coala.main(debug=True)`, the https://github.com/coala/bearsh is now able to:

```
In [2]: %PycodestyleBear setup.py
Executing section Default...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\Zimmermann\Miniconda3\envs\coala\lib\multiprocessing\spawn.py", line 106, in spawn_main
    exitcode = _main(fd)
  File "C:\Users\Zimmermann\Miniconda3\envs\coala\lib\multiprocessing\spawn.py", line 116, in _main
    self = pickle.load(from_parent)
EOFError: Ran out of input
---------------------------------------------------------------------------
PicklingError                             Traceback (most recent call last)
<ipython-input-2-a1efc0afb543> in <module>()
----> 1 get_ipython().magic('PycodestyleBear setup.py')

C:\Users\Zimmermann\Miniconda3\envs\coala\lib\site-packages\IPython\core\interactiveshell.py in magic(self, arg_s)
   2156         magic_name, _, magic_arg_s = arg_s.partition(' ')
   2157         magic_name = magic_name.lstrip(prefilter.ESC_MAGIC)
-> 2158         return self.run_line_magic(magic_name, magic_arg_s)
   2159
   2160     #-------------------------------------------------------------------------

C:\Users\Zimmermann\Miniconda3\envs\coala\lib\site-packages\IPython\core\interactiveshell.py in run_line_magic(self, magic_name, line)
   2077                 kwargs['local_ns'] = sys._getframe(stack_depth).f_locals
   2078             with self.builtin_trap:
-> 2079                 result = fn(*args,**kwargs)
   2080             return result
   2081

C:\Users\Zimmermann\Projects\coala\shell\bearsh\__init__.py in magic(arg_str, _bearname)
     36                             ','.join(arg_str.split())]
     37             try:
---> 38                 main(debug=True)
     39             finally:
     40                 sys.argv = _argv

c:\users\zimmermann\projects\coala\lib\coalib\coala.py in main(debug)
     85         return mode_format(debug=debug)
     86
---> 87     return mode_normal(console_printer, log_printer, args, debug=debug)
     88
     89 if __name__ == '__main__':  # pragma: no cover

c:\users\zimmermann\projects\coala\lib\coalib\coala_modes.py in mode_normal(console_printer, log_printer, args, debug)
     17         console_printer=console_printer,
     18         args=args,
---> 19         debug=debug)
     20
     21     return exitcode

c:\users\zimmermann\projects\coala\lib\coalib\coala_main.py in run_coala(console_printer, log_printer, print_results, acquire_settings, print_section_beginning, nothing_done, force_show_patch, arg_parser, arg_list, args, debug)
    108                 cache=cache,
    109                 log_printer=log_printer,
--> 110                 console_printer=console_printer)
    111             yielded, yielded_unfixed, results[section_name] = (
    112                 simplify_section_result(section_result))

c:\users\zimmermann\projects\coala\lib\coalib\processes\Processing.py in execute_section(section, global_bear_list, local_bear_list, print_results, cache, log_printer, console_printer)
    699
    700     for runner in processes:
--> 701         runner.start()
    702
    703     try:

C:\Users\Zimmermann\Miniconda3\envs\coala\lib\multiprocessing\process.py in start(self)
    103                'daemonic processes are not allowed to have children'
    104         _cleanup()
--> 105         self._popen = self._Popen(self)
    106         self._sentinel = self._popen.sentinel
    107         _children.add(self)

C:\Users\Zimmermann\Miniconda3\envs\coala\lib\multiprocessing\context.py in _Popen(process_obj)
    210     @staticmethod
    211     def _Popen(process_obj):
--> 212         return _default_context.get_context().Process._Popen(process_obj)
    213
    214 class DefaultContext(BaseContext):

C:\Users\Zimmermann\Miniconda3\envs\coala\lib\multiprocessing\context.py in _Popen(process_obj)
    311         def _Popen(process_obj):
    312             from .popen_spawn_win32 import Popen
--> 313             return Popen(process_obj)
    314
    315     class SpawnContext(BaseContext):

C:\Users\Zimmermann\Miniconda3\envs\coala\lib\multiprocessing\popen_spawn_win32.py in __init__(self, process_obj)
     64             try:
     65                 reduction.dump(prep_data, to_child)
---> 66                 reduction.dump(process_obj, to_child)
     67             finally:
     68                 context.set_spawning_popen(None)

C:\Users\Zimmermann\Miniconda3\envs\coala\lib\multiprocessing\reduction.py in dump(obj, file, protocol)
     57 def dump(obj, file, protocol=None):
     58     '''Replacement for pickle.dump() using ForkingPickler.'''
---> 59     ForkingPickler(file, protocol).dump(obj)
     60
     61 #

PicklingError: Can't pickle <PycodestyleBear linter class (wrapping 'pycodestyle')>: attribute lookup PycodestyleBear on coalib.bearlib.abstractions.Linter failed

In [3]: %debug
> c:\users\zimmermann\miniconda3\envs\coala\lib\multiprocessing\reduction.py(59)dump()
     57 def dump(obj, file, protocol=None):
     58     '''Replacement for pickle.dump() using ForkingPickler.'''
---> 59     ForkingPickler(file, protocol).dump(obj)
     60
     61 #

ipdb>
```

Which is also related to https://github.com/coala/coala/issues/3283

### CODE REVIEW

**Comment by sils** [comment]
@userzimmermann you got a conflict so marking WIP again, please resolve this.

@jayvdb I feel like you might want to have a say in this; generally it seems sane to have something like this and I also feel like it's ok to not have tests for it because it's basically just for us and not for our users. What's your take on that?

File: coalib/coala_main.py [hunk 1]
```diff
@@ -28,7 +28,9 @@ def run_coala(console_printer=None,
               autoapply=True,
               force_show_patch=False,
               arg_parser=None,
-              arg_list=None):
+              arg_list=None,
+              args=None,
+              debug=False):
```
Comment (NiklasMM) [inline coalib/coala_main.py]: `debug` needs a param section in the docstring.

File: requirements.txt [hunk 2]
```diff
@@ -2,6 +2,7 @@ appdirs~=1.4
 coala_utils~=0.6.2
 colorlog~=2.7
 dependency_management~=0.3.1
+# ipdb
```
Comment (NiklasMM) [inline requirements.txt]: why is this commented out?

Comment (userzimmermann) [inline requirements.txt]: @NiklasMM IIRC the `ipdb` requirement caused problems in CI. So I wanted to make it a dynamic `PipRequirement` when needed. That's why I also `import ipdb` only on demand. I will reproduce the actual problem again and paste it here...

**Comment by userzimmermann** [comment]
@NiklasMM Thanks for your review :)

The cli `--debug` arg is for the case when you just run `coala` from your system shell and want to get a debug prompt when something goes wrong. Just like `py.test --pdb`

The `debug` arg for `coalib.coala.main` and subsequent functions is for other tools that use `coala` programmatically, like the `bearsh`, so that any exceptions are just coming back uncatched and can be debugged afterwards, like with IPython's `%debug` magic in the `bearsh`

**Comment by userzimmermann** [comment]
@sils @NiklasMM @jayvdb I extended the whole thing with a `DebugProcessing` replacement for `multiprocessing`, which just bypasses any multiprocessing and directly runs the bears in the same Python process and thread. Also, the debug mode is now further passed to bear execution. So even every exception from inside bears' `.run` methods is now coming out of a coala run without getting catched before.

There is only one problem left: I want the debug mode to implicitly set log level to `DEBUG`, like when running `coala -V`. I tried it that way:

```python
args = default_arg_parser().parse_args()
if debug or args.debug:  # pragma: no cover
    args.log_level = 'DEBUG'
```

https://github.com/coala/coala/pull/3338/files#diff-aed82cdd504caf613603b250c092f460R33

But that doesn't seem to work. Is the `-V` flag itself processed in some other way?

**Comment by userzimmermann** [comment]
@sils @NiklasMM @jayvdb When adding a first `test_debug_find_no_issues` at https://github.com/coala/coala/pull/3338/files#diff-a67363bfc6d63f94e5926cf582cfa646R51 I realized an internal bug in `Processing.py`:

```
>       section_yielded_result = section_result[0]
E       TypeError: 'NoneType' object is not subscriptable

coalib\processes\Processing.py:651: TypeError
```

And after fixing this I found another one in `coala_main.py`:

```
>               file_dicts[section_name] = section_result[3]
E               TypeError: 'NoneType' object is not subscriptable

coalib\coala_main.py:128: TypeError
```

Those whole mechanisms too much rely on huge `try: ... except: ...` blocks with not really caring what actually happens in between!

Both fixes are now included in this PR. And all tests still pass!

**Comment by userzimmermann** [comment]
@sils @NiklasMM @jayvdb @Makman2 Added debug variants for every `coalaCITest`. And all of them pass :)

There is only one strange thing: 3 existing tests from other modules checked for `'During execution of coala'` in `stderr`:

`coalaTest.test_coala`

https://github.com/coala/coala/pull/3338/files#diff-bd7252a781a8e8027a95eb403d8d4118L37

`CachingTest.test_caching_results`

https://github.com/coala/coala/pull/3338/files#diff-f4c53753be8ec5d77fa199b20ac1caa4L129
https://github.com/coala/coala/pull/3338/files#diff-f4c53753be8ec5d77fa199b20ac1caa4L138

`CachingTest.test_caching_multi_results`

https://github.com/coala/coala/pull/3338/files#diff-f4c53753be8ec5d77fa199b20ac1caa4L165

And that error output just results from https://github.com/coala/coala/pull/3338/files#diff-f4c53753be8ec5d77fa199b20ac1caa4L165 , where it just wasn't checked if any results exist. Which I fixed in this PR. Now of course that error outputs don't happen anymore in those tests above. So I commented out the according assertions.

It seems like someone put those assertions in those tests without thinking about if that makes sense, and without investigating where those errors actually come from. What do you think?

**Comment by userzimmermann** [comment]
@sils If you are referring to `coala_ci`, `coala_format`, and `coala_json`:

I added the `debug` arguments to them only for consistency with my changes to `TestUtilities.execute_coala` :) Otherwise I would need to add some extra special handling for them...

File: coalib/coala_main.py [hunk 3]
```diff
@@ -60,6 +62,10 @@ def run_coala(console_printer=None,
     :param arg_parser:              Instance of ArgParser that is used to parse
                                     non-setting arguments.
     :param arg_list:                The CLI argument list.
+    :param args:                    The parsed CLI arguments.
```
Comment (sils) [inline coalib/coala_main.py]: what is the args behaviour if this isn't given (none)? IMO docs should cover that

File: coalib/parsing/DefaultArgParser.py [hunk 4]
```diff
@@ -230,6 +230,11 @@ def default_arg_parser(formatter_class=None):
         '-n', '--no-orig', const=True, action='store_const',
         help="don't create .orig backup files before patching")
 
+    misc_group.add_argument(
+        '--debug', const=True, action='store_const',
+        help='run coala in debug mode, starting ipdb on errors '
```
Comment (sils) [inline coalib/parsing/DefaultArgParser.py]: s/errors/exceptions/

File: coalib/processes/Processing.py [hunk 5]
```diff
@@ -366,13 +370,17 @@ def instantiate_processes(section,
     complete_filename_list = filename_list
     complete_file_dict = get_file_dict(complete_filename_list, log_printer)
 
-    manager = multiprocessing.Manager()
-    global_bear_queue = multiprocessing.Queue()
-    filename_queue = multiprocessing.Queue()
+    if debug:  # pragma: no cover
+        from . import DebugProcessing as processing
+    else:
+        processing = multiprocessing
```
Comment (sils) [inline coalib/processes/Processing.py]: remove import from top as discussed on telegram

File: tests/TestUtilities.py [hunk 6]
```diff
@@ -6,22 +6,33 @@
...
     sys.argv = [binary] + list(args)
     with retrieve_stdout() as stdout:
         with retrieve_stderr() as stderr:
-            retval = func()
+            try:
+                retval = func(debug=debug)
+            except Exception:
+                if debug:
+                    exc = sys.exc_info()
+                raise
```
Comment (sils) [inline tests/TestUtilities.py]: took me a while to see what you're doing here, I wonder if it could be simplified...

File: tests/coalaTest.py [hunk 7]
```diff
@@ -34,7 +34,7 @@ def test_coala(self):
             self.assertIn('This file has 1 lines.',
                           stdout,
                           'The output should report count as 1 lines')
-            self.assertIn('During execution of coala', stderr)
+            # self.assertIn('During execution of coala', stderr)
```
Comment (sils) [inline tests/coalaTest.py]: commented code is bad. Why?

Comment (userzimmermann) [inline tests/coalaTest.py]: @sils As described in https://github.com/coala/coala/pull/3338#issuecomment-293491125 , I first wanted to get some opinions about those asserts. If they have a useful purpose or if they were just introduced because there was some error output which did not really make sense but hey... before doing investigation better just accept and assert it ;)

I prefer just to remove those asserts, because those errors just arose from https://github.com/coala/coala/pull/3338/files#diff-f4c53753be8ec5d77fa199b20ac1caa4L165 and don't really make sense in the use cases of those tests.

File: tests/TestUtilities.py (hunk 6 above)
Comment (userzimmermann) [inline tests/TestUtilities.py]: I wonder too... :D The problem are the `with` contexts... But I will think about simplification

File: requirements.txt (hunk 2 above)
Comment (sils) [inline requirements.txt]: @userzimmermann maybe add some text like `# ipdb optional for debugging` so it gets clear

File: tests/misc/CachingTest.py [hunk 8]
```diff
@@ -126,7 +126,7 @@ def test_caching_results(self):
                 '-f', re.escape(filename),
                 '-b', 'LineCountTestBear')
             self.assertIn('This file has', stdout)
-            self.assertIn(' During execution of coala', stderr)
+            # self.assertIn(' During execution of coala', stderr)
```
Comment (sils) [inline tests/misc/CachingTest.py]: can you fix those errors in an own commit explaining how you fix them and in that commit also remove those messages? I think this should be orthogonal to what you do, this commit is way too big

Comment (userzimmermann) [inline tests/misc/CachingTest.py]: @sils Obsolete (for now) :D I made a little mistake :/ Will explain in separate comment...

**Comment by userzimmermann** [comment]
@sils @NiklasMM @jayvdb @Makman2 Oops! Little shame on me ;) My assumption in https://github.com/coala/coala/pull/3338#issuecomment-293491125 was a bit wrong. Those issue arose from a little mistake I made in the beginning of this PR, which prevented some results getting out of the processing internals.

Anyway: I meanwhile found out that the errors in those 3 tests leading to `'During execution of coala...'` in `stderr` come from missing input in `stdin`. I'm unsure if that is intended. It could be avoided by adding additional `with simulate_console_inputs(...)` contexts.

File: tests/coalaTest.py (hunk 7 above)
Comment (userzimmermann) [inline tests/coalaTest.py]: As already mentioned in https://github.com/coala/coala/pull/3338#discussion_r111240835 and further explained in https://github.com/coala/coala/pull/3338#issuecomment-293686866 , this is obsolete :)

File: coalib/parsing/DefaultArgParser.py (hunk 4 above)
Comment (userzimmermann) [inline coalib/parsing/DefaultArgParser.py]: Done :)

File: coalib/processes/Processing.py (hunk 5 above)
Comment (userzimmermann) [inline coalib/processes/Processing.py]: Done :)

File: requirements.txt (hunk 2 above)
Comment (userzimmermann) [inline requirements.txt]: @sils I directly took your suggestion :)

File: coalib/coala_main.py (hunk 1 above)
Comment (userzimmermann) [inline coalib/coala_main.py]: @NiklasMM Done :)

**Comment by sils** [comment]
@userzimmermann still lots of pragmas (and a small conflict); it looks good otherwise on first sight

File: coalib/coala.py [hunk 9]
```diff
@@ -69,15 +72,24 @@ def main():
             return 0
 
     except BaseException as exception:  # pylint: disable=broad-except
+        if not isinstance(exception, SystemExit):
+            if debug:  # pragma: no cover
```
Comment (userzimmermann) [inline coalib/coala.py]: @sils Any idea of how to write some proper fail test that raises something else than `SystemExit` here? To get the `# pragma` away

Comment (sils) [inline coalib/coala.py]: mock mode_json to just raise it

**Comment by userzimmermann** [comment]
@sils @jayvdb @Makman2 @sims1253  All `# pragma`s gone due to additional tests with some mocking :) Thanks @sils for the mocking suggestion!

ack ack ack :D

File: coalib/parsing/DefaultArgParser.py [hunk 10]
```diff
@@ -230,6 +230,11 @@ def default_arg_parser(formatter_class=None):
         '-n', '--no-orig', const=True, action='store_const',
         help="don't create .orig backup files before patching")
 
+    misc_group.add_argument(
+        '--debug', const=True, action='store_const',
+        help='run coala in debug mode, starting ipdb on internal exceptions '
+             '(implies --verbose)')
```
Comment (sils) [inline coalib/parsing/DefaultArgParser.py]: maybe we should mention that the user has to install ipdb manually

Comment (userzimmermann) [inline coalib/parsing/DefaultArgParser.py]: And we could use a `PipRequirement` when `--debug` flag is given to assert that `ipdb` is installed and step out and print install command if not

Comment (userzimmermann) [inline coalib/parsing/DefaultArgParser.py]: @sils Done!

**Comment by userzimmermann** [comment]
@sils WTF? Any idea? https://travis-ci.org/coala/coala/jobs/231285868#L328

```
error in ipdb setup command: Invalid environment marker: python_version >= "3.3"
```

Or maybe @jayvdb? You have such a talent for solving CI issues ;)

**Comment by userzimmermann** [comment]
Now I have missing coverage again by using that `PipRequirement('ipdb')` check :/

Maybe I should also mock that :D 

cc @sils

**Comment by userzimmermann** [comment]
@sils DONE 🎉

File: coalib/coala_main.py [hunk 11]
```diff
@@ -60,15 +62,23 @@ def run_coala(console_printer=None,
...
+    :param args:                    Alternative pre-parsed CLI arguments.
+    :param debug:                   Run in debug mode, bypassing
+                                    multiprocessing, and not catching any
+                                    exceptions.
     :return:                        A dictionary containing a list of results
                                     for all analyzed sections as key.
     """
+    assert not (arg_list and args), (
+        'Either provide run_coala() with an arg_list of CLI arguments or '
+        'pre-parsed args, but not both.')
```
Comment (sils) [inline coalib/coala_main.py]: can you put this check into gather_configuration instead? This only gets passed through here

File: coalib/output/Interactions.py [hunk 12]
```diff
@@ -23,4 +23,4 @@ def fail_acquire_settings(log_printer, settings_names_dict, section):
             msg += '{} (from {}) - {}'.format(name, setting[1], setting[0])
 
         log_printer.err(msg)
-        raise AssertionError
+        raise AssertionError(msg)
```
Comment (sils) [inline coalib/output/Interactions.py]: this change could be an own commit making the main commit a bit more slim

**Comment by sils** [comment]
btw. @Makman2 you might want to take a quick look at this we'll need that in the new core as well, turning off multiprocessing for coala debugging

**Comment by Makman2** [comment]
Turning off multiprocessing in the next-gen core is as easy as cake: I already plan to let `core.run` to take an `executor` parameter. Then you can just supply a `ThreadPoolExecutor` when running the core :+1:

**Comment by sils** [comment]
@Makman2 no not switching to threads, actually running it synchronous

**Comment by Makman2** [comment]
Then we have to write our own executor, Python sadly doesn't have one^^ Shouldn't be that hard though^^

**Comment by Makman2** [comment]
though you sure a `ThreadPoolExecutor` won't work?

File: coalib/coala_main.py [hunk 13]
```diff
@@ -135,6 +144,17 @@ def run_coala(console_printer=None,
         elif yielded_results:
             exitcode = 5
     except BaseException as exception:  # pylint: disable=broad-except
+        if not isinstance(exception, SystemExit):
+            if args and args.debug or sections and getattr(
+                    sections.get('cli', {}).get('debug'), 'value', False
```
Comment (sils) [inline coalib/coala_main.py]: I think this needs a bool conversion

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/3338/commits/fa4abdd13e92c33b85df699c54030d0e44a75e66 after many many in-person iterations

**Comment by userzimmermann** [comment]
@rultor merge
